### PR TITLE
Added DiVA fetcher based on the new interface

### PR DESCRIPTION
**Pull request by oscargus**
- [x] Tests created for changes

### CODE REVIEW

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        return HelpFile.FETCHER_DIVA_TO_BIBTEX;
+    }
+
+    @Override
+    public Optional<BibEntry> performSearchById(String identifier) throws FetcherException {
+        String q;
+        try {
+            q = URLEncoder.encode(identifier, StandardCharsets.UTF_8.name());
+        } catch (UnsupportedEncodingException e) {
+            // this should never happen
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Just throw a FetcherException as you do below.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import net.sf.jabref.logic.importer.FetcherException;
+import net.sf.jabref.logic.importer.IdBasedFetcher;
+import net.sf.jabref.logic.net.URLDownload;
+import net.sf.jabref.model.entry.BibEntry;
+import net.sf.jabref.model.entry.FieldName;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class DiVA implements IdBasedFetcher {
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Can shortly explain what DiVA is in a comment and add a link to its API.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        String q;
+        try {
+            q = URLEncoder.encode(identifier, StandardCharsets.UTF_8.name());
+        } catch (UnsupportedEncodingException e) {
+            // this should never happen
+            LOGGER.warn("Encoding issues", e);
+            return Optional.empty();
+        }
+
+        String urlString = String.format(DiVA.URL_PATTERN, q);
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Use Apache`s URIBuilder to create the URL.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    @Override
+    public String getName() {
+        return "DiVA";
+    }
+
+    @Override
+    public HelpFile getHelpPage() {
+        return HelpFile.FETCHER_DIVA_TO_BIBTEX;
+    }
+
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Could you please also add a method `isValidId(String identifier)` which checks that a given ID is of the expected format. I plan to add some functionality which depends on such a method but have no insights what DiVA identifiers look like. Thanks

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            URLDownload dl = new URLDownload(url);
+
+            bibtexString = dl.downloadToString(StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            throw new FetcherException("Communication problems", e);
+        }
+
+        Optional<BibEntry> result = Optional.ofNullable(BibtexParser.singleFromString(bibtexString));
+
+        result.ifPresent(entry -> entry.getFieldOptional(FieldName.INSTITUTION).ifPresent(institution -> entry
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: I would leave it to the users cleanup actions to format unicode to latex.

File: src/test/java/net/sf/jabref/logic/importer/fetcher/DiVATest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+
+public class DiVATest {
+
+    private DiVA fetcher;
+
+
+    @Before
+    public void setUp() {
+        Globals.prefs = JabRefPreferences.getInstance();
```
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/logic/importer/fetcher/DiVATest.java]: Prefs are not required, right?

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java (hunk 2 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: I suggest to add this fetcher to the list of web searches and also remove the old one (or is there reason to keep the old code?)

File: src/test/java/net/sf/jabref/logic/importer/fetcher/DiVATest.java (hunk 6 above)
Comment (oscargus) [inline src/test/java/net/sf/jabref/logic/importer/fetcher/DiVATest.java]: They are required to parse the BibTeX entry (wrapped fields).

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java (hunk 2 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: As there is no easy way to wrap an IdBased searcher at the moment, it is
not possible. There was a class in the ISBN stupro PR so once that is merge
it makes sense.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java (hunk 4 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: I have no idea what the requirements are. Most I've seen are on the format `diva2:XXXXX` though, where X is a digit.

Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: I tried to play around with the API and it appears that the identifier definitely has to start with `diva2` so just take this as a test for valid identifier.

**Comment by oscargus** [comment]
@JabRef/developers: I rebased this and introduced a change in `MetaData`, making constructors without provided `Charset` deprecated. This is an attempt to get rid of the `Globals` dependency. Didn't change anything else related to that in this PR, but once merged there are many things to be changed.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public Optional<BibEntry> performSearchById(String identifier) throws FetcherException {
+        Optional<BibEntry> result = Optional.empty();
+        try {
+            URIBuilder uriBuilder = new URIBuilder(URL);
+
+            uriBuilder.addParameter("referenceFormat", "BibTex");
+            uriBuilder.addParameter("pids", identifier);
+
+            String bibtexString;
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Combine this line with line 55.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            URLDownload dl = new URLDownload(uriBuilder.build().toURL());
+
+            bibtexString = dl.downloadToString(StandardCharsets.UTF_8);
+            result = Optional.ofNullable(BibtexParser.singleFromString(bibtexString, importFormatPreferences));
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Just `return Optional.of(...)` and remove result variable.

**Comment by tobiasdiez** [comment]
Some small remarks. After fixing these and the codacy issue, this can be merged.

File: src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java (hunk 7 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/importer/fetcher/DiVA.java]: Yeah, weird code... :-)

**Pull request merged**